mlflow/evaluation.py

```diff
@@ -0,0 +1,354 @@
+        if self.run_id is not None:
+            return self.run_id
+        else:
+            self.managed_run_context = mlflow.start_run()
+            return self.managed_run_context.__enter__().info.run_id
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.managed_run_context is not None:
+            return self.managed_run_context.__exit__(exc_type, exc_val, exc_tb)
+
+
+class ModelEvaluator:
+
+    def can_evaluate(
+        self, model_type, evaluator_config=None, **kwargs
+    ) -> bool:
+        """
+        :param model_type: A string describing the model type (e.g., "regressor",
+                           "classifier", …).
+        :param evaluator_config: A dictionary of additional configurations for
+                                 the evaluator.
+        :param **kwargs: For forwards compatibility, a placeholder for additional
+                         arguments that may be added to the evaluation interface
+                         in the future.
+        :return: True if the evaluator can evaluate the specified model on the
+                 specified dataset. False otherwise.
+        """
+        raise NotImplementedError()
+
+    def compute_metrics(self, predict, dataset):
+        """
+        return an instance of EvaluationMetrics
+        """
+        raise NotImplementedError()
+
+    def compute_and_log_artifacts(self, predict, dataset, run_id, mlflow_client):
+        """
+        compute and log artifact, and return a dict of
+        artifact_name -> instance_of_EvaluationArtifacts
+        """
+        raise NotImplementedError()
+
+    def evaluate(
+        self, predict, dataset, run_id=None, evaluator_config=None, **kwargs
+    ) -> EvaluationResult:
+        """
+        :param predict: A function used to compute model predictions. Predict
+                        accepts features from the specified `dataset` and
+                        feeds them to the model, producing output predictions.
+        :param dataset: An instance of `EvaluationDataset` containing features
+                        and labels (optional) for model evaluation.
+        :param run_id: The ID of the MLflow Run to which to log results.
+        :param evaluator_config: A dictionary of additional configurations for
+                                 the evaluator.
+        :param **kwargs: For forwards compatibility, a placeholder for additional
+                         arguments that may be added to the evaluation interface
+                         in the future.
+        :return: An `EvaluationResult` instance containing evaluation results.
+        """
+        client = mlflow.tracking.MlflowClient()
+        with GetOrCreateRunId(run_id) as run_id:
+            metrics_dict = self.compute_metrics(predict, dataset)
+            timestamp = int(time.time() * 1000)
+            dataset_id = dataset.name if dataset.name is not None else dataset.hash
+            # TODO: log tags of dataset metadata
```

Comment:
In design doc, the tags format is a list of dataset metadata
But implementing it make performance bad. When we append a new dataset metadata into the tag, we need first read tag `mlflow.datasets`, then append the new metadata into the tag string, and then overwrite the tag.

shall we change the design to log tag key like `mlflow.datasets.{name/hash}` and with each dataset metadata value ?

Reply:
@WeichenXu123 We need to make sure that users can search for runs based on dataset names, hashes, and paths. We don't have partial search capabilities for tag keys, so it might be tricky to make this work. Let me know if you have thoughts about supporting search here.

In general, we don't expect that users will call `mlflow.evaluate()` frequently enough for the performance to matter. Concurrency is a concern, but the likelihood of a race condition seems low; we can handle it with locking later.